File: scripts/readers/skills.py
"""Reader for Claude Code skills (`.claude/skills/<name>/SKILL.md`)."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from scripts.util import split_frontmatter


def _collect_extra_files(skill_dir: Path) -> list[Path]:
    """Collect auxiliary files in the skill directory (excluding SKILL.md).

    Returns paths relative to `skill_dir`. Directories are walked recursively;
    only files are returned.
    """
    extras: list[Path] = []
    for entry in sorted(skill_dir.iterdir()):
        if entry.name == "SKILL.md":
            continue
        if entry.is_file():
            extras.append(entry.relative_to(skill_dir))
        elif entry.is_dir():
            for sub in sorted(entry.rglob("*")):
                if sub.is_file():
                    extras.append(sub.relative_to(skill_dir))
    return extras


def read_skills(project_root: Path) -> list[dict[str, Any]]:
    """Walk `<project_root>/.claude/skills/*/SKILL.md` and return skill records.

    Each record is a dict with keys: id, name, description, when_to_use,
    frontmatter, body, extra_files.

    Skill directories that lack a SKILL.md are skipped gracefully. If
    `.claude/skills/` does not exist, an empty list is returned.
    """
    skills_root = project_root / ".claude" / "skills"
    if not skills_root.is_dir():
        return []

    records: list[dict[str, Any]] = []
    for skill_dir in sorted(p for p in skills_root.iterdir() if p.is_dir()):
        skill_md = skill_dir / "SKILL.md"
        if not skill_md.is_file():
            continue

        text = skill_md.read_text(encoding="utf-8")
        frontmatter, body = split_frontmatter(text)

        skill_id = skill_dir.name
        name = frontmatter.get("name") or skill_id
        description = frontmatter.get("description") or ""
        when_to_use = frontmatter.get("when_to_use")

        records.append(
            {
                "id": skill_id,
                "name": name,
                "description": description,
                "when_to_use": when_to_use,
                "frontmatter": frontmatter,
                "body": body,
                "extra_files": _collect_extra_files(skill_dir),
            }
        )

    return records

File: plugins/cc-codex-bridge/src/readers/skills.ts
import fs from "node:fs";
import path from "node:path";

import { splitFrontmatter, type Frontmatter } from "../util.js";

export interface SkillRecord {
  id: string;
  name: string;
  description: string;
  when_to_use: string | null;
  frontmatter: Frontmatter;
  body: string;
  extra_files: string[];
}

function collectExtraFiles(skillDir: string): string[] {
  const extras: string[] = [];
  const entries = fs.readdirSync(skillDir, { withFileTypes: true });
  entries.sort((a, b) => a.name.localeCompare(b.name));
  for (const entry of entries) {
    if (entry.name === "SKILL.md") continue;
    const full = path.join(skillDir, entry.name);
    if (entry.isFile()) {
      extras.push(entry.name);
    } else if (entry.isDirectory()) {
      walk(full, skillDir, extras);
    }
  }
  return extras;
}

function walk(dir: string, root: string, out: string[]): void {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  entries.sort((a, b) => a.name.localeCompare(b.name));
  for (const entry of entries) {
    const full = path.join(dir, entry.name);
    if (entry.isFile()) {
      out.push(path.relative(root, full).split(path.sep).join("/"));
    } else if (entry.isDirectory()) {
      walk(full, root, out);
    }
  }
}

export function readSkills(projectRoot: string): SkillRecord[] {
  const skillsRoot = path.join(projectRoot, ".claude", "skills");
  if (!isDirectory(skillsRoot)) {
    return [];
  }

  const records: SkillRecord[] = [];
  const skillDirs = fs
    .readdirSync(skillsRoot, { withFileTypes: true })
    .filter((e) => e.isDirectory())
    .map((e) => e.name)
    .sort();

  for (const name of skillDirs) {
    const skillDir = path.join(skillsRoot, name);
    const skillMd = path.join(skillDir, "SKILL.md");
    if (!isFile(skillMd)) continue;

    const text = fs.readFileSync(skillMd, "utf-8");
    const [frontmatter, body] = splitFrontmatter(text);

    const nameField = typeof frontmatter.name === "string" && frontmatter.name.length > 0
      ? frontmatter.name
      : name;
    const description = typeof frontmatter.description === "string" ? frontmatter.description : "";
    const whenToUse = typeof frontmatter.when_to_use === "string" ? frontmatter.when_to_use : null;

    records.push({
      id: name,
      name: nameField,
      description,
      when_to_use: whenToUse,
      frontmatter,
      body,
      extra_files: collectExtraFiles(skillDir),
    });
  }

  return records;
}

function isDirectory(p: string): boolean {
  try {
    return fs.statSync(p).isDirectory();
  } catch {
    return false;
  }
}

function isFile(p: string): boolean {
  try {
    return fs.statSync(p).isFile();
  } catch {
    return false;
  }
}
